pkg/spiffe/certfetcher.go

```diff
@@ -0,0 +1,126 @@
+// Package spiffe contains code related to spiffe identity management
+package spiffe
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var errNoClient = fmt.Errorf("failed to start cert fetcher: nil client")
+
+//go:generate counterfeiter -generate
+
+// Client wraps the workloadapi.Client
+//
+//counterfeiter:generate ./. Client
+type Client interface {
+	WatchX509Context(context.Context, workloadapi.X509ContextWatcher) error
+	Close() error
+}
+
+// implements workloadapi.x509ContextWatcher.
+type watcher struct {
+	certFetcher X509CertFetcher
+}
+
+func newWatcher(fetcher X509CertFetcher) *watcher {
+	return &watcher{certFetcher: fetcher}
+}
+
+// OnX509ContextUpdate is called when a new X.509 Context is fetched from the SPIFFE Workload API.
+// The X.509 context is placed on the cert fetcher's cert channel.
+func (w *watcher) OnX509ContextUpdate(svidResp *workloadapi.X509Context) {
+	log.Printf("SVID updated for spiffeID: %q\n", svidResp.DefaultSVID().ID)
+	w.certFetcher.CertCh <- svidResp
+}
+
+// OnX509WatchError is called when there is an error watching the X.509 Context's from the SPIFFE Workload API.
+func (w *watcher) OnX509ContextWatchError(err error) {
+	msg := "For more information check the logs of the Spire agents and server."
+	switch status.Code(err) { //nolint:exhaustive // uses a default
+	case codes.Unavailable:
+		log.Printf("X509SVIDClient cannot connect to the Spire agent: %v. %s", err, msg)
+	case codes.PermissionDenied:
+		log.Printf("X509SVIDClient still waiting for certificates: %v. %s", err, msg)
+	case codes.Canceled:
+		return
+	default:
+		log.Printf("X509SVIDClient error: %v. %s", err, msg)
+	}
+}
+
+// CertFetcher fetches certificates
+//
+//counterfeiter:generate ./. CertFetcher
+type CertFetcher interface {
+	// Start starts fetching X.509 certificates.
+	// It returns an error if it fails to start.
+	// Otherwise, certificates are written to the X509Context channel
+	// and if there is an unrecoverable error it is written to the error channel.
+	Start(context.Context) (<-chan *workloadapi.X509Context, <-chan error, error)
+	// Stop closes the connection with the SPIFFE Workload API Client.
+	Stop() error
+}
+
+// X509CertFetcher fetches certs from the X509 SPIFFE Workload API.
+type X509CertFetcher struct {
+	client     Client
+	WatchErrCh chan error
+	CertCh     chan *workloadapi.X509Context
+	spireAddr  string
+}
+
+// NewX509CertFetcher creates a new instance of CertFetcher.
+func NewX509CertFetcher(spireAddr string, client Client) (*X509CertFetcher, error) {
```

Comment:
What about the Options pattern? See https://www.sohamkamani.com/golang/options-pattern/

I think that would eliminate the need for anyone to pass `nil` to the constructor.

Reply:
Resolved out of band to consider the options pattern in the future for more complex modules.